### ISSUE TITLE
Same idea with StyleGAN-Encoder repo?

### ISSUE THREAD

**Issue by sherock369** [genforce/idinvert#1]
Thanks for your work. The motivation is really inspiring, but I have two questions:
1. The pipeline of your work is really similar to [StyleGAN-Encoder.](https://github.com/pbaylies/stylegan-encoder) Can you explain the difference?

2. In-domain is an insightful idea, but I cannot get how you invert the image to the semantic-adaptive domain. Using the domain-guided encoder? However, I think that it can only push the inverted image to be realism enough but is unable to ensure the semantic-adaptive characteristic since no explicit or implicit semantic-related constraints are applied. In other words, it indeed conducts the same thing with previous works, i.e., Image2StyleGAN.

**Comment by ShenYujun** [comment]
1. There are two major differences, i.e. _domain-guided encoder_ and _domain-regularized optimization_. (a) `stylegan-encoder` does not actually have an encoder to map a given image back to the latent space. By contrast, our encoder provides a much better initialization. (b) We also involve the encoder in the optimization process as a regularizer, to ensure the code to be in-domain. BTW, `Image2StyleGAN`, instead of this repo, is much like `stylegan-encoder` (They use exactly the same process, and the only difference is the loss function). As for your mentioned "conducting the same thing", I think all GAN inversion approaches do the same thing, which is to recover the input image with latent code(s). Differently, we raise the problem of exploring the semantic property of the inverted codes. This distinguishes our work from existing work.

2. We explain this in the paper. You are right that it is the _domain-guided encoder_ to ensure the semantic property. The intuitive explanation is the training manner of the encoder. During the encoder training, we will feed the output of the encoder back into the generator to synthesize a photo-realistic image. In this way, if the code does not land inside the original latent space, the synthesis quality will suffer. Alternatively, the code should align with the semantic knowledge learned by the generator for image synthesis. That is what we call _in-domain_. Fig.3 in the paper also demonstrates this point. To maintain the code to be semantically meaningful, we also involve the encoder in the instance-level optimization. The [demo video](https://www.youtube.com/watch?v=3v6NHrhuyFY) shows the superiority of our approach over the optimization-based method (like `stylegan-encoder` and `Image2StyleGAN`). With the help of the _domain-guided encoder_, the inverted code is more manipulable.

**Comment by ShenYujun** [comment]
BTW, we have [another work](https://genforce.github.io/mganprior/) I think you may have interests in. That work proposes a different optimization-based pipeline by composing multiple latent codes for a single image.

**Comment by sherock369** [comment]
Thanks! I have read all your works about StyleGAN-inversing/ editing. I have got the idea of your work. First, learning an encoder to inverse the image to the latent space of image2stylegan. Second, since the encoder is not representative enough, you add a refinement step to obtain more accurate results. This idea is in line with the trend of the community. 

1. Maybe you ignore something in StyleGAN-Encoder. There are many popular variants of StyleGAN-Encoder. What I point is a modified repo, which trains an encoder to provide initialization for optimization. It sounds more principled that the first step is explained as learning semantic properties, and there are some slight differences between yours and StyleGAN-Encoder. But the pipeline is actually the same. Please refer to [this](https://github.com/pbaylies/stylegan-encoder) for more details. And I strongly recommend you read their valuable discussions on issues.

2. I have got what you mean semantic preserving.  The encoder can preserve some properties but not ALL. I have tried exactly the same architecture with only difference in loss weights. In my experiment, some properties are easy to inverse, such as pose, expression, but some special properties are extremely hard to inverse, such as hairstyle, eye color, and glasses.  My implementation is based on Pytorch, and the code of the StyleGAN generator is from InterfaceGAN. I think this point is potential and could be dug deeper. Any signs of progress will inform you.

In summary, some insightful explanations and experiments in your work are principled and interesting.  Nice work.

**Comment by sherock369** [comment]
The training data you use is relatively clean. Maybe this point affects the effect of the domain-guided encoder. I will check it again.

**Comment by ShenYujun** [comment]
Thanks for your "strong recommendation"! We never claim our pipeline, which first uses an encoder to provide initialization and them performs optimization for an individual image, is novel. This is a standard pipeline in the GAN inversion task (we also mention some references in Sec.2). However, we improve both steps (please also note that the encoder is also involved in the optimization step instead of just providing initialization) and study the semantic property of the inverted code, which is never explored before. Existing work typically focuses on pixel-wise reconstruction, yet we argue that semantic meaning should be another metric to evaluate an inversion approach.

Also, by comparing the results from the encoder in the [repo](https://github.com/pbaylies/stylegan-encoder) you mentioned and those in our papers, I still think our domain-guided encoder is novel and better. For example, our encoder can produce a more satisfying reconstruction before optimization. We also include a comparison with the traditional encoder in Fig.4 of our paper. Hence, I reaffirm that we use a standard pipeline, which almost all inversion approaches follow, but we improve both steps. And please carefully find _all the claims, details, and experiments_ provided in our paper if you are interested. Good luck!

**Comment by sherock369** [comment]
Yes, you are right! Pixel-level constraints are harmful to StyleGAN-inversing. Very principled work.

**Comment by sherock369** [comment]
Hi Yujun, after carefully reading your paper and much consideration and experiments, I find some differences in our conclusions.

1. For training the domain-guided encoder, the perceptual loss is enough to keep the output keeping semantical. Using MSE or adv loss is redundant and not as well as perceptual loss. Using `L_E` has no little influence on the encoder. Maybe you can double-check. 

2. You argue that existing work freely optimizes the latent code, which may lead to an out-of-domain latent code. But if you have checked the issuse1 or the code in StyleGANencoder, you will find they already use regular loss or truncation trick to prevent this point.....

Frankly speaking, the stylegan-encoder has a weaker encoder, but they consider the optimization step more comprehensively. The stylegan-encoder indeed is the SOTA for StyleGAN inversing task, but you didn't compare your work with them. What a pity!

**Comment by sherock369** [comment]
1. In fact, you train on real images instead of synthesized images by introducing an adversarial manner. Am I right? In your paper and as you mentioned above, you find perceptual loss is more effective. So why you claim the choice of loss isn't your contribution. The first improvement is trivial and is used in previous work. StyleGANencoder also trains the encoder by projecting the code to an image. I just focus on the nature of the problem. No offense to your work and your contribution. Instead, I think this paper has some inspiring points.

2. The discussion of regular loss and truncation trick is in [issue2](https://github.com/pbaylies/stylegan-encoder/issues/2).  Sorry for this mistake.

3. Yes, speed is the bottleneck for gan inverse task.

4. If you try to test the stylegan-encoder, you will find it works really well. I'm just curious about the comparison between your work and the stylegan-encoder, which is not necessary for your paper.

Anyway, I didn't want to share my experiment results publicly. So I will delete all comments once you receive my reply. Welcome to communicate with me by email (sherock.cu@outlook.com).
